### ISSUE TITLE
[Fine-tuning for other languages] How to obtain or train "qwen-tts-tokenizer" used in Qwen2.5-Omni?

### ISSUE THREAD

**Issue by SeungyounShin** [QwenLM/Qwen2.5-Omni#219]
I'm currently exploring the **Qwen2.5-Omni** multimodal model (as described in the [Qwen2.5-Omni Technical Report](https://arxiv.org/abs/2503.20215) and particularly interested in adapting it to generate speech in languages beyond English and Chinese, specifically **Korean**.

The Qwen2.5-Omni paper describes that the **Talker module** generates speech using discrete codec tokens produced by an audio tokenizer named **"qwen-tts-tokenizer"**:

> “We designed an efficient speech codec named **qwen-tts-tokenizer**. **qwen-tts-tokenizer** efficiently represents key information of speech and can be decoded to speech streamingly through a causal audio decoder.”

However, the publicly available implementation (e.g., on Hugging Face) appears to only include the decoder part of the model pipeline:

- **Talker**: Text → speech codec tokens  
- **Token2Wav**: Speech codec tokens → mel-spectrogram → waveform  

I couldn't find any publicly accessible method or code snippet for the reverse step (**waveform → codec tokens**) using the mentioned **"qwen-tts-tokenizer"**.

This leads to a few questions:

1. **Is the "qwen-tts-tokenizer" publicly available or open-sourced?**  
   If yes, could you point me to the repository or implementation?

2. If not publicly available, **are there plans to release the tokenizer or its pretrained checkpoints?**

3. Alternatively, **could you share what kind of model architecture or codec method (e.g., EnCodec, SoundStream, VQ-VAE) the tokenizer uses?**  
   This information would be extremely helpful to recreate or fine-tune a similar tokenizer.

Having access to this tokenizer is critical to training the **Talker** module for new languages (like Korean), as otherwise, creating the necessary training data (**text ↔ codec token pairs**) is not possible.

I appreciate any guidance, details, or alternative recommendations you can provide.

Thank you!

**Comment by wulaoshi** [comment]
Same problem.

**Comment by abrar360** [comment]
I'm looking into this as well and since the Qwen3-omni technical report has also been released now, I'm wondering if this has any additional clues since certain parts of the pipeline may be similar. This paper reports that they are now "operating directly on RVQ tokens", which may suggest a similar approach to Descript Audio Codec using RVQGAN.

I wonder if the following approach to recreate a similar tokenizer might work:

1. Generate TTS (text, codec tokens, waveform) pairs by prompting the model "Say the following: [phrase]".
2. Use these (codec tokens, waveform) pairs to train an RVQGAN encoder.

EDIT: It seems very similar to what this blogpost for Sesame CSM is describing (https://www.sesame.com/research/crossing_the_uncanny_valley_of_voice).

**Comment by hashiting** [comment]
@abrar360 @SeungyounShin Hi, any news here？

**Comment by hashiting** [comment]
> I'm looking into this as well and since the Qwen3-omni technical report has also been released now, I'm wondering if this has any additional clues since certain parts of the pipeline may be similar. This paper reports that they are now "operating directly on RVQ tokens", which may suggest a similar approach to Descript Audio Codec using RVQGAN.
> 
> I wonder if the following approach to recreate a similar tokenizer might work:
> 
> 1. Generate TTS (text, codec tokens, waveform) pairs by prompting the model "Say the following: [phrase]".
> 2. Use these (codec tokens, waveform) pairs to train an RVQGAN encoder.
> 
> EDIT: It seems very similar to what this blogpost for Sesame CSM is describing (https://www.sesame.com/research/crossing_the_uncanny_valley_of_voice).

@abrar360 hi, this might work with the same speaker tune. How to clone a new voice?

**Comment by abrar360** [comment]
@hashiting Here's an experiment I was looking to try, but I haven't had the time:

1. Take Qwen3-omni and add logging to huggingface transformers to store the RVQ codes that are ingested by the Code2Wav.
2. Run Qwen3-omni and use it to generate audio as a WAV file, the raw RVQ codes should be logged.
3. Run the audio through the Mimi encoder (https://github.com/kyutai-labs/moshi) and compare these with the RVQ codes we logged.
4. Take the logged RVQ codes and feed them through the Mimi decoder and see if you can play the audio.

If this works, then that means that Qwen is directly built on the Mimi audio codec just like Sesame-CSM!

Now all you have to do to clone a new voice is to get your voice cloning audio data and encode it with Mimi and use this to fine-tune only the MTP module. (The MTP module takes in semantic codebook tokens from the Talker, and outputs acoustic tokens).

**Comment by SeungyounShin** [comment]
I did what @abrar360 said and it was right.

1. breakpoint() before this [line](https://github.com/huggingface/transformers/blob/bdee0889714e9cb3e53d3b1b2a626919479d356c/src/transformers/models/qwen3_omni_moe/modular_qwen3_omni_moe.py#L2588)

2. log RVQ code as numpy (1,16,L)
3. Run 
```python
from datasets import load_dataset, Audio
from transformers import MimiModel, AutoFeatureExtractor
import numpy as np
import soundfile as sf
from pathlib import Path
import torch

ASSETS_DIR = Path("Qwen3-Omni/assets")

# load the model
model = MimiModel.from_pretrained("kyutai/mimi")

# load the feature extractor
feature_extractor = AutoFeatureExtractor.from_pretrained("kyutai/mimi")

# load preprocessed audio codes
audio_codes = np.load(ASSETS_DIR / "talker_codes.npy")
audio_codes = torch.tensor(audio_codes)

# decode the audio codes
audio_values = model.decode(audio_codes.to(model.device))

import pdb; pdb.set_trace()

# save the audio
sf.write("mimi.wav", audio_values.audio_values.squeeze().detach().cpu().numpy(), feature_extractor.sampling_rate)

```

4. Take the logged RVQ codes and feed them through the Mimi decoder and see if you can play the audio.

Got same result from qwen3-omni and mimi reconstructed data

[qwen_output.wav](https://github.com/user-attachments/files/23701605/qwen_output.wav)
[mimi_recon.wav](https://github.com/user-attachments/files/23701606/mimi_recon.wav)

**Comment by hashiting** [comment]
> I did what [@abrar360](https://github.com/abrar360) said and it was right.
> 
> 1. breakpoint() before this [line](https://github.com/huggingface/transformers/blob/bdee0889714e9cb3e53d3b1b2a626919479d356c/src/transformers/models/qwen3_omni_moe/modular_qwen3_omni_moe.py#L2588)
> 2. log RVQ code as numpy (1,16,L)
> 3. Run
> 
> from datasets import load_dataset, Audio
> from transformers import MimiModel, AutoFeatureExtractor
> import numpy as np
> import soundfile as sf
> from pathlib import Path
> import torch
> 
> ASSETS_DIR = Path("Qwen3-Omni/assets")
> 
> # load the model
> model = MimiModel.from_pretrained("kyutai/mimi")
> 
> # load the feature extractor
> feature_extractor = AutoFeatureExtractor.from_pretrained("kyutai/mimi")
> 
> # load preprocessed audio codes
> audio_codes = np.load(ASSETS_DIR / "talker_codes.npy")
> audio_codes = torch.tensor(audio_codes)
> 
> # decode the audio codes
> audio_values = model.decode(audio_codes.to(model.device))
> 
> import pdb; pdb.set_trace()
> 
> # save the audio
> sf.write("mimi.wav", audio_values.audio_values.squeeze().detach().cpu().numpy(), feature_extractor.sampling_rate)
> 4. Take the logged RVQ codes and feed them through the Mimi decoder and see if you can play the audio.
> 
> Got same result from qwen3-omni and mimi reconstructed data
> 
> [qwen_output.wav](https://github.com/user-attachments/files/23701605/qwen_output.wav) [mimi_recon.wav](https://github.com/user-attachments/files/23701606/mimi_recon.wav)

@SeungyounShin @abrar360 thx a lot for sharing, your guys must be genius. One more question, do you guys have the demo script or sth to train the talker part?

**Comment by abrar360** [comment]
@SeungyounShin That's huge! Thanks for sharing your results!

I'm glad it ended up working. 

It looks like Qwen-3-omni supports Korean, are you still looking to add a new language?

**Comment by abrar360** [comment]
@hashiting I don't have any scripts yet for this.

**Comment by wulaoshi** [comment]
> # load the feature extractor
> feature_extractor = AutoFeatureExtractor.from_pretrained("kyutai/mimi")



> I did what [@abrar360](https://github.com/abrar360) said and it was right.我做了说的，而且对了。
> 
> 1. breakpoint() before this [line](https://github.com/huggingface/transformers/blob/bdee0889714e9cb3e53d3b1b2a626919479d356c/src/transformers/models/qwen3_omni_moe/modular_qwen3_omni_moe.py#L2588)在这行之前调试断点
> 2. log RVQ code as numpy (1,16,L)将 RVQ 代码以 numpy 格式记录 （1,16，L）
> 3. Run  运行
> 
> from datasets import load_dataset, Audio
> from transformers import MimiModel, AutoFeatureExtractor
> import numpy as np
> import soundfile as sf
> from pathlib import Path
> import torch
> 
> ASSETS_DIR = Path("Qwen3-Omni/assets")
> 
> # load the model
> model = MimiModel.from_pretrained("kyutai/mimi")
> 
> # load the feature extractor
> feature_extractor = AutoFeatureExtractor.from_pretrained("kyutai/mimi")
> 
> # load preprocessed audio codes
> audio_codes = np.load(ASSETS_DIR / "talker_codes.npy")
> audio_codes = torch.tensor(audio_codes)
> 
> # decode the audio codes
> audio_values = model.decode(audio_codes.to(model.device))
> 
> import pdb; pdb.set_trace()
> 
> # save the audio
> sf.write("mimi.wav", audio_values.audio_values.squeeze().detach().cpu().numpy(), feature_extractor.sampling_rate)
> 4. Take the logged RVQ codes and feed them through the Mimi decoder and see if you can play the audio.取所记录的 RVQ 码，通过 Mimi 解码器处理，看看是否能播放音频。
> 
> Got same result from qwen3-omni and mimi reconstructed data从 qwen3-omni 和 mimi 重建的数据中得到相同的结果
> 
> [qwen_output.wav](https://github.com/user-attachments/files/23701605/qwen_output.wav) [mimi_recon.wav](https://github.com/user-attachments/files/23701606/mimi_recon.wav)



> I did what [@abrar360](https://github.com/abrar360) said and it was right.
> 
> 1. breakpoint() before this [line](https://github.com/huggingface/transformers/blob/bdee0889714e9cb3e53d3b1b2a626919479d356c/src/transformers/models/qwen3_omni_moe/modular_qwen3_omni_moe.py#L2588)
> 2. log RVQ code as numpy (1,16,L)
> 3. Run
> 
> from datasets import load_dataset, Audio
> from transformers import MimiModel, AutoFeatureExtractor
> import numpy as np
> import soundfile as sf
> from pathlib import Path
> import torch
> 
> ASSETS_DIR = Path("Qwen3-Omni/assets")
> 
> # load the model
> model = MimiModel.from_pretrained("kyutai/mimi")
> 
> # load the feature extractor
> feature_extractor = AutoFeatureExtractor.from_pretrained("kyutai/mimi")
> 
> # load preprocessed audio codes
> audio_codes = np.load(ASSETS_DIR / "talker_codes.npy")
> audio_codes = torch.tensor(audio_codes)
> 
> # decode the audio codes
> audio_values = model.decode(audio_codes.to(model.device))
> 
> import pdb; pdb.set_trace()
> 
> # save the audio
> sf.write("mimi.wav", audio_values.audio_values.squeeze().detach().cpu().numpy(), feature_extractor.sampling_rate)
> 4. Take the logged RVQ codes and feed them through the Mimi decoder and see if you can play the audio.
> 
> Got same result from qwen3-omni and mimi reconstructed data
> 
> [qwen_output.wav](https://github.com/user-attachments/files/23701605/qwen_output.wav) [mimi_recon.wav](https://github.com/user-attachments/files/23701606/mimi_recon.wav)

From your experimental results, is it possible to infer that the Code2Wav module in Qwen3 Omni is actually the Mimi decoder?

**Comment by abrar360** [comment]
@wulaoshi Although they seem functionally similar (both convert Mimi RVQ codes to wav), there seem to be key architectural differences:

Namely, the Mimi decoder incorporates Transformer layers:

<img width="1313" height="457" alt="Image" src="https://github.com/user-attachments/assets/ff2d139d-6c8a-4154-ade3-d29140c6b564" />

On the other hand, Code2Wav is said to be "a lightweight causal ConvNet" in the technical report.

**Comment by wulaoshi** [comment]
> [@wulaoshi](https://github.com/wulaoshi) Although they seem functionally similar (both convert Mimi RVQ codes to wav), there seem to be key architectural differences:
> 
> Namely, the Mimi decoder incorporates Transformer layers:
> 
> <img alt="Image" width="1313" height="457" src="https://private-user-images.githubusercontent.com/20074831/523561968-ff2d139d-6c8a-4154-ade3-d29140c6b564.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjUzNDM4NzYsIm5iZiI6MTc2NTM0MzU3NiwicGF0aCI6Ii8yMDA3NDgzMS81MjM1NjE5NjgtZmYyZDEzOWQtNmM4YS00MTU0LWFkZTMtZDI5MTQwYzZiNTY0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjEwVDA1MTI1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwMjJmODJjYTZiZjI2ZTI5YjRhMjE4MmM3OWQzNzU4NDkzNDIyOTYyY2UwNDU2ZGVjZmZkNjJhODI5NDk4OGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.n0JRyhrbSMhY-ILlTADrk1F7g-MQfp9c8r4NEE20C-M">
> On the other hand, Code2Wav is said to be "a lightweight causal ConvNet" in the technical report.

Yes, according to the technical report, they are indeed different.

**Comment by JoshonSmith** [comment]
So far, has anyone managed to train the talker part?
